components/utilities/floatingNav.js

```diff
@@ -1,204 +1,158 @@
 
-    this.generateMenu();
-  }
+    const newNestedHeadings = getNestedHeadings(headingElements);
+    setNestedHeadings(newNestedHeadings);
+  }, []);
 
-  handleRouteChange() {
-    this.closeMenu();
-    this.generateMenu();
-  }
+  return { nestedHeadings };
+};
 
-  componentWillUnmount() {
-    this.closeMenu();
-    window.removeEventListener("ChangeTheme", this.handleTheme);
-    this.props.router.events.off("routeChangeComplete", this.handleRouteChange);
-  }
+const getNestedHeadings = (headingElements) => {
+  const nestedHeadings = [];
 
-  handleTheme() {
-    this.setState({ theme: document.body.dataset.theme });
+  for (const index in headingElements) {
+    const ele = headingElements[index];
+    if (ele.getElementsByTagName === undefined) {
+      continue;
+    }
+    const hrefs = ele.getElementsByTagName("a");
+    if (hrefs.length > 0) {
+      const target = hrefs[0].getAttribute("href");
+      nestedHeadings.push({
+        label: ele.innerText,
+        target: target,
+        level: ele.tagName,
+      });
+    }
   }
 
-  render() {
-    let color;
-    let svgColor = this.state.theme === "light-mode" ? "black" : "white";
-    const props = this.props;
-    const menu = this.state.menu;
-    const target = this.state.target;
-    const isnum = /^[\d\.]+$/.test(props.slug[0]);
-    const slug = isnum ? props.slug.slice(1).join("/") : this.state.slug;
-    const location = this.state.slug.split("/")[0];
-
-    // Get the Root Object and find the appropriate color
-    const breadCrumbs = breadcrumbsForSlug(props.menu, `/${slug}`);
-
-    if (breadCrumbs.length) {
-      const rootElement = breadCrumbs[0];
-      const darkColor = rootElement.color;
-      color = darkColor;
-      if (this.state.theme === "light-mode") {
-        const colorParts = darkColor.split("-");
-        let lightMode = colorParts.length > 1 ? Number(colorParts[1]) - 50 : 10;
-        if (lightMode < 10 || isNaN(lightMode)) lightMode = 10;
-        color = `${colorParts.slice(0, -1).join("-")}-${lightMode}`;
-        svgColor = darkColor;
+  return nestedHeadings;
+};
+
+// This is the function that initializes the intersection observer, and attaches it to the elements we want to track, our page headings
+const useIntersectionObserver = (setActiveId) => {
+  // Since we want to keep track of the visibility of all the heading elements, we'll store these values in a useRef hook
+  const headingElementsRef = useRef({});
+
+  useEffect(() => {
+    // This is our callback function. The observer will call this function each time elements scroll in or out of view.
+    // When we first render the page, it calls the callback with a list of all the elements we want to keep track on the page. As elements scroll in and out of view, it will call the callback with these elements to update the visibility.
+    const callback = (headings) => {
+      headingElementsRef.current = headings.reduce((map, headingElement) => {
```

Comment:
I think this would be cleaner/more understandable as a simple forEach:

```js
      headings.forEach((headingElement) => {
        const link = headingElement.target.getElementsByTagName("a");
        if (link.length > 0) {
          const target = link[0].getAttribute("href");
          headingElementsRef.current[target] = headingElement;
        }
      });
```

Reply:
Fixed on d27b9aa